toolchain/check/context.cpp

```diff
@@ -339,45 +336,49 @@ static auto ProfileType(Context& semantics_context, SemIR::Node node,
                         llvm::FoldingSetNodeID& canonical_id) -> void {
   switch (node.kind()) {
     case SemIR::NodeKind::ArrayType: {
-      auto [bound_id, element_type_id] = node.GetAsArrayType();
+      auto array_type = node.As<SemIR::ArrayType>();
       canonical_id.AddInteger(
-          semantics_context.semantics_ir().GetArrayBoundValue(bound_id));
-      canonical_id.AddInteger(element_type_id.index);
+          semantics_context.semantics_ir().GetArrayBoundValue(
+              array_type.bound_id));
+      canonical_id.AddInteger(array_type.element_type_id.index);
```

Comment:
In general for these cases where there is only one node really at play over a short scope with the explicit type already visible, I don't find losing the structured binding to be a net readability improvement... I guess the concern is keeping the destructured names in sync with the member names?

I somewhat feel like the destructuring brevity is maybe reasonable (and can be kept even w/ the new API) in super brief contexts like this?

Not arguing to keep it always though, especially across multiple scopes or with multiple different nodes where it's useful to make their accesses distinct and type checked.

Reply:
Generally my concern with destructuring is that the names could get out of sync, or the field order could be accidentally reversed. The same problem exists when creating the nodes, though, and this patch isn't addressing it, so maybe that's a somewhat weak argument?  (On the other hand, we could support providing designators when creating nodes, if we wanted to make that side more explicit too.)

There's also a future-looking aspect here: @josh11b had mentioned a desire to switch to something more visitor-ish for these switches over node kinds, replacing

```c++
 switch (node.kind()) {
   case SemIR::NodeKind::ArrayType: {
      auto array_type = node.As<SemIR::ArrayType>();
      canonical_id.AddInteger(
          semantics_context.semantics_ir().GetArrayBoundValue(
              array_type.bound_id));
      canonical_id.AddInteger(array_type.element_type_id.index);
      break;
    }
    case SemIR::NodeKind::Builtin:
      canonical_id.AddInteger(node.As<SemIR::Builtin>().builtin_kind.AsInt());
      break;
    // ...
```

with something like

```c++
 node.Visit(
   [&] (SemIR::ArrayType array_type) {
      canonical_id.AddInteger(
          semantics_context.semantics_ir().GetArrayBoundValue(
              array_type.bound_id));
      canonical_id.AddInteger(array_type.element_type_id.index);
    },
    [&] (SemIR::Builtin builtin) {
      canonical_id.AddInteger(builtin.builtin_kind.AsInt());
    },
    // ...
```

and switching away from structured bindings towards the typed nodes prepares us for that.

Reply:
The absence of even a clang-tidy check to keep destructured names consistent, and the relative difficulty of having *internal* type safety within them is persuading me we should just use members as you have here.

However, I think the visitor thing needs (much) more discussion. I don't think the language support for this in C++ is good enough to be usable -- the lambda pattern I think will make compile time and other things scale really badly over time. Happy to chat about options here though, and that seems like a separate change, if and where we want to make it.

Reply:
I think I was suggesting a visitor for parse nodes. Anyway, I don't think there is any changes for this PR.